autodocodec-api-usage/src/Autodocodec/Usage.hs

```diff
@@ -493,3 +494,90 @@ instance HasCodec MultilineDefault where
     object "MultilineDefault" $
       MultilineDefault
         <$> optionalFieldWithDefault "value" (Via "foo" "bar") "a field with a multi-line default value" .= multilineDefaultValue
+
+data These
+  = This Text
+  | That Int
+  | Both Text Int
+  deriving stock (Show, Eq, Generic)
+  deriving
+    ( FromJSON,
+      ToJSON,
+      Swagger.ToSchema,
+      OpenAPI.ToSchema
+    )
+    via (Autodocodec These)
+
+instance Validity These
+
+instance NFData These
+
+instance GenValid These where
+  genValid = genValidStructurallyWithoutExtraChecking
+  shrinkValid = shrinkValidStructurallyWithoutExtraFiltering
+
+instance HasCodec These where
+  codec =
+    object "These" $
+      discriminatedUnionCodec "type" enc dec
+    where
+      textFieldCodec = requiredField' "text"
+      intFieldCodec = requiredField' "int"
+      bothFieldsCodec = (,) <$> textFieldCodec .= fst <*> intFieldCodec .= snd
+      enc = \case
+        This s -> ("this", mapToEncoder s textFieldCodec)
+        That n -> ("that", mapToEncoder n intFieldCodec)
+        Both s n -> ("both", mapToEncoder (s, n) bothFieldsCodec)
+      dec =
+        HashMap.fromList
+          [ ("this", ("This", mapToDecoder This textFieldCodec)),
+            ("that", ("That", mapToDecoder That intFieldCodec)),
+            ("both", ("Both", mapToDecoder (uncurry Both) bothFieldsCodec))
+          ]
+
+data Expression
+  = LiteralExpression Int
+  | SumExpression Expression Expression
+  | ProductExpression Expression Expression
+  deriving stock (Show, Eq, Generic)
+  deriving
+    ( FromJSON,
+      ToJSON,
+      Swagger.ToSchema,
+      OpenAPI.ToSchema
+    )
+    via (Autodocodec Expression)
+
+instance Validity Expression where
+  validate = trivialValidation
+
+instance NFData Expression
+
+instance GenValid Expression where
+  genValid = sized $ \size -> do
+    if size > 0
+      then
+        oneof
+          [ LiteralExpression <$> genValid,
+            SumExpression <$> scale (`div` 2) genValid <*> scale (`div` 2) genValid,
+            ProductExpression <$> scale (`div` 2) genValid <*> scale (`div` 2) genValid
+          ]
```

Comment:
Nit: This is good enough here, but you probably would have wanted to use `genSplit` to split the size instead.

Reply:
I changed it to use `genSplit`. I wasn't aware of that.